beginner_source/blitz/tensor_tutorial.py

```diff
@@ -1,195 +1,189 @@
-print(x.size(), y.size(), z.size())
+print(f"Shape of tensor: {tnsr.shape}")
+print(f"Datatype of tensor: {tnsr.dtype}")
+print(f"Device tensor lives on: {tnsr.device}")
 
-###############################################################
-# If you have a one element tensor, use ``.item()`` to get the value as a
-# Python number
-x = torch.randn(1)
-print(x)
-print(x.item())
 
-###############################################################
-# **Read later:**
-#
+######################################################################
+# --------------
 #
-#   100+ Tensor operations, including transposing, indexing, slicing,
-#   mathematical operations, linear algebra, random numbers, etc.,
-#   are described
-#   `here <https://pytorch.org/docs/torch>`_.
-#
-# NumPy Bridge
-# ------------
+
+
+######################################################################
+# Tensor Operations
+# ~~~~~~~~~~~~~~~~~
 #
-# Converting a Torch Tensor to a NumPy array and vice versa is a breeze.
+# Over 100 tensor operations, including transposing, indexing, slicing,
+# mathematical operations, linear algebra, random sampling, and more are
+# comprehensively described
+# `here <https://pytorch.org/docs/stable/torch.html>`__.
 #
-# The Torch Tensor and NumPy array will share their underlying memory
-# locations (if the Torch Tensor is on CPU), and changing one will change
-# the other.
+# Each of them can be run on the GPU (at typically higher speeds than on a
+# CPU). If you’re using Colab, allocate a GPU by going to Edit > Notebook
+# Settings.
 #
-# Converting a Torch Tensor to a NumPy Array
-# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-a = torch.ones(5)
-print(a)
+# We move our tensor to the GPU if available
+if torch.cuda.is_available():
+  tnsr = tnsr.to('cuda')
+
+
+######################################################################
+# Try out some of the operations from the list. If you're familiar with the NumPy API, you'll find the Tensor API a breeze to use.
+#
 
 ###############################################################
+# **Standard numpy-like indexing and slicing:**
+
+tnsr = torch.ones(4, 4)
+tnsr[:,1] = 0
+print(tnsr)
+
+######################################################################
+# **Both of these are joining ops, but they are subtly different.**
+
+t1 = torch.cat([tnsr, tnsr], dim=1)
+t2 = torch.stack([tnsr, tnsr], dim=1)
+print(t1)
+print()
+print(t2)
+
+######################################################################
+# **Multiply op - multiple syntaxes**
+
+# both ops compute element-wise product
+print(tnsr * tnsr == tnsr.mul(tnsr))
+
+# both ops compute matrix product
+print(tnsr @ tnsr.T == tnsr.matmul(tnsr.T))
+
+######################################################################
+# .. note::
+#     Operations that have a '_' suffix are in-place. For example:
+#     ``x.copy_(y)``, ``x.t_()``, will change ``x``. In-place operations
+#     don't work well with Autograd, and their use is discouraged.
```

Comment:
In-place operations save some memory, but they don't always work to compute derivative formulas because of immediate loss-of-history. Hence, their use is discouraged.